betterjails/src/main/java/io/github/emilyydev/betterjails/interfaces/storage/StorageInterface.java

```diff
@@ -0,0 +1,22 @@
+package io.github.emilyydev.betterjails.interfaces.storage;
+
+import com.github.fefo.betterjails.api.model.jail.Jail;
+import io.github.emilyydev.betterjails.api.impl.model.prisoner.ApiPrisoner;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+public interface StorageInterface {
+  CompletableFuture<Void> savePrisoner(ApiPrisoner prisoner);
+  CompletableFuture<Void> deletePrisoner(ApiPrisoner prisoner);
+  CompletableFuture<Void> loadPrisoners(Map<UUID, ApiPrisoner> out);
+
+  CompletableFuture<Void> saveJails(Map<String, Jail> jails);
+  CompletableFuture<Void> loadJails(Map<String, Jail> out);
+
+  // TODO(rymiel): There's asymmetry between these: "save one, delete one, load all" vs "save all, load all"
+  //   there's also some extra asymmetry between ApiPrisoner vs Jail, this is because not all data needed here
+  //   is stored in Prisoner. Perhaps some stuff from ApiPrisoner should be exposed in Prisoner.
+  //   also, not sure if out parameters are a good idea.
```

Comment:
I think the interface should have methods to deal with individual jails as well, and then it's on the implementation to figure out how to pull it off. I think it's reasonable to have a `saveJail` method as well a `deleteJail` method on the interface, it's just an unfortunate situation that the way jails are stored with the Bukkit Configuration API doesn't really allow for it, but it shouldn't be the one dictating how this interface is modelled. `savePrisoners` maybe? Still have to think whether `saveJails` should save the given jails only or delete all the ones not in the map as well, the former sounds the most sensible but we shall see.

The reasoning behind the out parameters is partial results when it comes to failures, so now that brings the question of what to do with the already loaded data if we get a partial result, should the result be merged or replace the existing data? Regardless of partial results, I don't think using an out parameter makes sense on a method that is by nature asynchronous (returns CompletableFuture), maybe have a Result interface with Failure/Success/PartialResult or something